Document Template system (#114)
Mirantis/hmc: docs/templates/main.md

# Templates system

By default, Hybrid Container Cloud delivers the set of default `Template` objects. It's also supported to build your

> own templates and use it for the deployment.

Reviewer: Let's mention, that templates can only reside in the hmc-system namespace at the moment, but can be referenced by Deployments from any namespace.

## Custom deployment Templates

> > At the moment all deployment Templates should reside in the `hmc-system` namespace. But it can be referenced

Reviewer: ```suggestion
> At the moment all `Templates` should reside in the `hmc-system` namespace. But they can be referenced
```

1. Create [HelmRepository](https://fluxcd.io/flux/components/source/helmrepositories/) object containing the URL to the
external helm repository.
2. Create [HelmChart](https://fluxcd.io/flux/components/source/helmcharts/) referencing the `HelmRepository` as a

> `sourceRef`, the name and version of a helm chart containing cluster objects.

Reviewer: ```suggestion
`sourceRef`, specifying the name and version of your Helm chart.
```

> 3. Create `Template` object referencing this helm chart in `spec.helm.chartRef`. `chartRef` is a field of

Reviewer: ```suggestion
3. Create a `Template` object referencing this helm chart in `spec.helm.chartRef`. `chartRef` is a field of the
```

[CrossNamespaceSourceReference](https://fluxcd.io/flux/components/helm/api/v2/#helm.toolkit.fluxcd.io/v2.CrossNamespaceSourceReference) kind.

The `Template` should follow the rules mentioned below:
1. `spec.type` should be `deployment` (as an alternative, the referenced helm chart may contain the
`hmc.mirantis.com/type: deployment` annotation in `Chart.yaml`).
2. `spec.providers` should contain the list of required Cluster API providers: `infrastructure`, `bootstrap` and
`controlPlane`. As an alternative, the referenced helm chart may contain the specific annotations in the
`Chart.yaml` (value is a list of providers divided by comma). These fields are only used for validation. For example:

`Template` spec:

```yaml
spec:
  providers:
    infrastructure:
      - aws
    bootstrap:
      - k0s
    controlPlane:
      - k0s
      - k0smotron
```

`Chart.yaml`:

```bash
annotations:
  hmc.mirantis.com/infrastructure-providers: aws
  hmc.mirantis.com/controlplane-providers: k0s,k0smotron
  hmc.mirantis.com/bootstrap-providers: k0s
```

> ### Example

Reviewer: Let's provide examples along the steps

Here is an example of a custom `Template` with the `HelmChart` reference:

```yaml
apiVersion: source.toolkit.fluxcd.io/v1
kind: HelmRepository
metadata:
  name: custom-templates-repo
  namespace: hmc-system
spec:
  insecure: true
  interval: 10m0s
  provider: generic
  type: oci
  url: oci://ghcr.io/external-templates-repo/charts
```

```yaml
apiVersion: source.toolkit.fluxcd.io/v1
kind: HelmChart
metadata:
  name: custom-template-chart
  namespace: hmc-system
spec:
  interval: 5m0s
  chart: custom-template-chart-name
  reconcileStrategy: ChartVersion
  sourceRef:
    kind: HelmRepository
    name: custom-templates-repo
  version: 0.2.0
```

```yaml
apiVersion: hmc.mirantis.com/v1alpha1
kind: Template
metadata:
  name: os-k0smotron
  namespace: hmc-system
spec:
  type: deployment
  providers:
    infrastructure:
      - openstack
    bootstrap:
      - k0s
    controlPlane:

>       - k0s

Reviewer: let's choose only one of those

      - k0smotron
  helm:
    chartRef:
      kind: HelmChart
      name: custom-template-chart
      namespace: default
```

## Remove Templates shipped with HMC

> If you need to limit cluster templates that exist in your HMC installation, follow the instruction below:

Reviewer: ```suggestion
If you need to limit the cluster templates that exist in your HMC installation, follow the instructions below:
```

1. Get the list of `deployment` Templates shipped with HMC:

```bash

> kubectl get template -n hmc-system -o go-template='{{ range .items }}{{ if and (eq .status.type "deployment") (eq (index .metadata.annotations "meta.helm.sh/release-name") "hmc-templates") }}{{ .metadata.name }}{{ printf "\n" }}{{ end }}{{ end }}'

Reviewer: let's simplify and use --field-selector. we only need satus.type == deployment
